### PR TITLE
Stop-on-fail verifier [blocks: 3968]

### PR DESCRIPTION
**Pull request by peterschrammel**
~Based on #3795, only review last 5 commits.~

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 175837625c887364a2968843590d741c4ec70a6f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97973645

File: src/goto-checker/stop_on_fail_verifier.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    switch(determine_result(properties))
+    {
+    case resultt::PASS:
+      report_success(ui_message_handler);
+      incremental_goto_checker.output_proof();
+      break;
+
+    case resultt::FAIL:
+    {
+      namespacet ns(goto_model.get_symbol_table());
```
Comment (tautschnig) [inline src/goto-checker/stop_on_fail_verifier.h]: `const`

File: src/goto-checker/stop_on_fail_verifier.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#define CPROVER_GOTO_CHECKER_STOP_ON_FAIL_VERIFIER_H
+
+#include "bmc_util.h"
+#include "goto_verifier.h"
+
+/// Stops when the first failing property is found.
+/// Requires an incremental goto checker that is a
+/// `goto_trace_providert` and `witness_providert`.
+template <class incremental_goto_checkerT>
+class stop_on_fail_verifiert : public goto_verifiert
```
Comment (tautschnig) [inline src/goto-checker/stop_on_fail_verifier.h:22]: I am wondering whether a `stop_on_fail_verifiert` genuinely _has_ an `incremental_goto_checkerT` or rather should _be_ an `incremental_goto_checkerT`?!

**Review by tautschnig — APPROVED** [review]

File: src/goto-checker/stop_on_fail_verifier.h (hunk 2 above)
Comment (peterschrammel) [reply src/goto-checker/stop_on_fail_verifier.h:22]: You'd still need a wrapper around the `incremental_goto_checkert` to do the result reporting.

File: src/goto-checker/stop_on_fail_verifier.h (hunk 1 above)
Comment (peterschrammel) [reply src/goto-checker/stop_on_fail_verifier.h]: Fixed

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4b022437876099481668a79d4fd54e1916ecf8cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98754158

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3b71199efa9d5be8f3e7daefc54e64aaf8631f71).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99353532

**Review by smowton — APPROVED** [review]

**Pull request merged**